elasticdl/python/common/tensor.py

```diff
@@ -113,3 +84,93 @@ def to_tf_tensor(self):
             return tf.IndexedSlices(self.values, self.indices)
         else:
             return tf.constant(self.values)
+
+    def to_ndarray(self):
+        if self.is_indexed_slices():
+            # Currently Tensor does not have a field representing dense shape,
+            # thus can not convert it to numpy.ndarray.
+            raise NotImplementedError(
+                "Converting an ElasticDL Tensor object, which contains a "
+                "sparse tensor, to a numpy.ndarray is not supported."
+            )
+        return self.values
+
+
+def serialize_tensor(tensor, tensor_pb):
+    """Serialize ElasticDL Tensor to tensor protocol buffer."""
+    dtype = dtype_numpy_to_tensor(tensor.values.dtype)
+    if not dtype:
+        raise ValueError(
+            "Dtype of ndarray %s is not supported", tensor.values.dtype
+        )
+    tensor_pb.dtype = dtype
+    tensor_pb.dim.extend(tensor.values.shape)
+    tensor_pb.content = tensor.values.tobytes()
+    if tensor.is_indexed_slices():
+        tensor_pb.indices.extend(tuple(tensor.indices))
+    if tensor.name:
+        tensor_pb.name = tensor.name
+
+
+def deserialize_tensor_pb(tensor_pb, tensor):
+    """Deserialize tensor protocol buffer to ElasticDL Tensor.
+
+    Note that the input tensor protocol buffer is reset and underlying buffer
+    is passed to the returned ndarray.
+    """
+    if not tensor_pb.dim:
+        raise ValueError("Tensor PB has no dim defined")
+
+    dtype = dtype_tensor_to_numpy(tensor_pb.dtype)
+    # Check that the buffer size agrees with dimensions.
+    size = dtype.itemsize
+    for d in tensor_pb.dim:
+        size *= d
+    if size != len(tensor_pb.content):
+        raise ValueError(
+            "Tensor PB size mismatch, dim: %s, len(content): %d",
+            tensor_pb.dim,
+            len(tensor_pb.content),
+        )
+    tensor.set(
+        values=np.ndarray(
+            shape=tensor_pb.dim, dtype=dtype, buffer=tensor_pb.content
+        ),
+        indices=np.array(tensor_pb.indices) if tensor_pb.indices else None,
+        name=tensor_pb.name,
+    )
+    tensor_pb.Clear()
```

Comment:
Please remove this line, `tensor_pb.Clear()`?

Reply:
And also please note that protobuf Clear method will not release the memory. It just set the fields of the message to default values, but keep in memory. Please refer to :

- https://github.com/protocolbuffers/protobuf/issues/6197
- https://itindex.net/detail/51781-protobuf-%E7%A8%8B%E5%BA%8F-%E5%86%85%E5%AD%98

Reply:
There are also some discussion in https://github.com/sql-machine-learning/elasticdl/pull/1361#discussion_r338644206. 

In the offline discussion, we have some different opinions. Will open an issue to discuss it and modify in another PR if needs.